### DISCUSSION TITLE
Disabling ASLR for binaries executed in the autograder disallowed

### DISCUSSION THREAD

**Discussion by dkohlbre** [gradescope/autograder_samples#124] [Q&A]
Hi,
I teach a security course where some of the assignments involve writing exploits for simple targets.
To make this easier on the students, we disable ASLR and enable exec stacks.
On our servers this is relatively easily done with calls to `personality` (e.g. `personality(ADDR_NO_RANDOMIZE |READ_IMPLIES_EXEC)`)

When I've tried to build an autograder for this assignment I can't make those calls from a binary running in the Docker container.
(Nor can you do things like modify the relevant `/proc/` bits, its mounted RO)
As far as I'm aware, this is due to the seccomp policy the container is run under. I don't believe you can modify that policy from the container image, only from the way the container is started.

Does anyone have a solution for disabling ASLR/execstack for specific binaries inside the autograder?

Thanks.

**Comment by cfm** [comment]
It looks like it would be possible to add support for these `personality()` calls in gVisor.[^1][^2]  There's been discussion about whether gVisor itself [requires some modification](https://github.com/google/gvisor/issues/4371) to run under Docker's default seccomp profile, but there's a more-recent production use-case documented in <https://github.com/freedomofpress/dangerzone/blob/main/docs/developer/gvisor.md>.

[^1]: https://github.com/google/gvisor/issues/153
[^2]: https://github.com/google/gvisor/issues/10756

**Comment by witchof0x20** [comment]
Also teaching a security course, also interested in this.

**Reply by dkohlbre** [reply]
Sadly I still don't have a solution for this.
The best I have is a half-finished and untested attempt at building a full qemu image I can run inside the container to accurately emulate the servers where students do their work.
Gradescope did talk to me via email, but they weren't able to come up with anything about changing their seccomp policies.

**Reply by witchof0x20** [reply]
> The best I have is a half-finished and untested attempt at building a full qemu image I can run inside the container to accurately emulate the servers where students do their work. 

Thanks for the tip on this. I have to upload a 1gb zip file every time I fix up the grader, but this actually worked!

**Reply by dkohlbre** [reply]
I haven't had time to look into it, so really glad to hear it did work!
@witchof0x20 Do you have a public build script for that I could work off of?

**Reply by witchof0x20** [reply]
Oops, I haven't looked at this issue in a while. Let me send you an email with some files.
